### PR TITLE
Convert hauser main process to service and add testing

### PR DESCRIPTION
**Pull request by camphillips22**
This change reworks the main hauser process to use dependency injection so
that the client interfaces can be mocked. I added some very basic
testing to prove that the standard process is working correctly.

I plan on following up with some more structural changes to make
the configuration behavior a little more consistent.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -26,12 +26,21 @@ func (e StatusError) Error() string {
...
 
+// DataExportClient represents an interface for interacting with the FullStory Data Export API
+type DataExportClient interface {
+	ExportList(start time.Time) ([]ExportMeta, error)
+	ExportData(id int, modifyReq ...func(r *http.Request)) (ExportData, error)
+}
+
 // Client represents a HTTP client for making requests to the FullStory API.
 type Client struct {
+	DataExportClient
```
Comment (claygoddard) [inline client/client.go:37]: I don't think you want this inlined here. The `Client` itself is the one that should be implementing the interface. The type check on line 42 has you covered. Plus, calling `Client.DataExportClient.ExportList()`, for example, would panic..

File: client/client.go [hunk 2]
```diff
@@ -26,12 +26,21 @@ func (e StatusError) Error() string {
 	return fmt.Sprintf("fullstory: response error: Status:%s, StatusCode:%d, RetryAfter:%v", e.Status, e.StatusCode, e.RetryAfter)
 }
 
+// DataExportClient represents an interface for interacting with the FullStory Data Export API
+type DataExportClient interface {
+	ExportList(start time.Time) ([]ExportMeta, error)
```
Comment (claygoddard) [inline client/client.go:31]: Quick method comments for the new interface?

File: client/client.go (hunk 1 above)
Comment (camphillips22) [reply client/client.go:37]: Good call. I meant to remove this.

File: config/config.go [hunk 3]
```diff
@@ -107,5 +108,12 @@ func Load(filename string) (*Config, error) {
 		conf.ExportURL = DefaultExportURL
 	}
 
+	if conf.TmpDir == "" {
+		conf.TmpDir, err = ioutil.TempDir("", "hauser")
```
Comment (claygoddard) [inline config/config.go:112]: Is this a safe change? What was happening for empty string folks before? Storing in the relative current directory?

File: main.go [hunk 4]
```diff
@@ -37,7 +35,20 @@ const (
...
 
-type ExportProcessor func(warehouse.Warehouse, []string, *client.Client, []client.ExportMeta) (int, error)
+type HauserService struct {
+	config       *config.Config
+	fsClient     client.DataExportClient
+	warehouse    warehouse.Warehouse
+	tableColumns []string
+}
+
+func NewHauser(config *config.Config, fsClient client.DataExportClient, warehouse warehouse.Warehouse) *HauserService {
```
Comment (claygoddard) [inline main.go:45]: While we are refactoring here, should we move `HauserService` out of `main.go` into its own file or package?

File: config/config.go (hunk 3 above)
Comment (camphillips22) [reply config/config.go:112]: Ah, it looks like it does just resolve to the current directory (makes sense 😄 ). I can remove this change.

File: main.go [hunk 5]
```diff
@@ -349,20 +368,55 @@ func BackoffOnError(err error) bool {
...
+	return h.ProcessExportsSince(lastSyncedRecord)
+}
+
+func (h *HauserService) Run() {
+	if err := h.Init(); err != nil {
+		log.Fatal(err)
+	}
+	for {
+		if numBundles, err := h.ProcessNext(); err != nil {
+			if h.BackoffOnError(err) {
```
Comment (claygoddard) [inline main.go:399]: I think you want to unconditionally call `BackoffOnError` here so that it resets when there is no error.

File: main_test.go [hunk 6]
```diff
@@ -1,17 +1,132 @@
...
 	"github.com/fullstorydev/hauser/client"
+	"github.com/fullstorydev/hauser/config"
+	hausertest "github.com/fullstorydev/hauser/testing"
 	"github.com/fullstorydev/hauser/warehouse"
 	"github.com/pkg/errors"
 )
 
+var update = flag.Bool("update", false, "update upload files")
+
+func Ok(t *testing.T, err error, format string, a ...interface{}) {
```
Comment (claygoddard) [inline main_test.go:25]: Should these go in some testing utils package?

**Comment by claygoddard** [comment]
A few comments, once addressed things should be good to go!

File: main.go (hunk 5 above)
Comment (camphillips22) [reply main.go:399]: Ah, yes. Of course!

File: main.go (hunk 4 above)
Comment (camphillips22) [reply main.go:45]: I can do that. Do you think we should move it to an `internal` package? Or just something like "core"?

Comment (claygoddard) [reply main.go:45]: `internal/service.go` seems reasonable to me!

**Pull request merged**